### ISSUE TITLE
When ack is enabled and consumer starts after publisher has published all messages, the consumer receives messages from one queue only

### ISSUE THREAD

**Issue by susam** [pika/pika#410]
I face this issue in the following two environments:

  1. pika 0.9.13, Python 2.7.3, RabbitMQ 3.1.5 and Windows 7 Enterprise Service Pack 1.
  2. pika 0.9.13, Python 2.6.6, RabbitMQ 3.0.1 and CentOS release 6.4.

Use the following two scripts to reproduce the issue.

### send.py ###

    import pika
    import logging

    logging.basicConfig(level=logging.DEBUG)

    parameters = pika.ConnectionParameters('localhost')
    connection = pika.BlockingConnection(parameters)
    channel = connection.channel()

    # Declare queues for the consumer
    channel.queue_declare(queue='q1')
    channel.queue_declare(queue='q2')

    # Publish messages
    for i in range(3):
        message = 'message ' + str(i)
        channel.basic_publish(exchange='',
                              routing_key='q1',
                              body='q1: ' + message)
        channel.basic_publish(exchange='',
                              routing_key='q2',
                              body='q2: ' + message)
        print 'Published', message, 'with routing keys q1 and q2'

    connection.close()

### recv.py ###

    import pika
    import logging

    logging.basicConfig(level=logging.DEBUG)

    def print_data(channel, method, properties, body):
        print 'body:', body
        channel.basic_ack(delivery_tag = method.delivery_tag)

    parameters = pika.ConnectionParameters('localhost')
    connection = pika.BlockingConnection(parameters)
    channel = connection.channel()

    channel.queue_declare(queue='q1')
    channel.queue_declare(queue='q2')

    channel.basic_consume(print_data, queue='q1')
    channel.basic_consume(print_data, queue='q2')

    channel.start_consuming()
    connection.close()

Now execute send.py first.

    C:\pika\one-queue-issue>python send.py 2> sendlog.txt
    Published message 0 with routing keys q1 and q2
    Published message 1 with routing keys q1 and q2
    Published message 2 with routing keys q1 and q2

    C:\pika\one-queue-issue>rabbitmqctl list_queues
    Listing queues ...
    q1      3
    q2      3
    ...done.

Now execute recv.py. It begins waiting after processing messages from q1. Press Ctrl + C to kill it.

    C:\pika\one-queue-issue>python recv.py 2> recvlog.txt
    body: q1: message 0
    body: q1: message 1
    body: q1: message 2

    C:\pika\one-queue-issue>rabbitmqctl list_queues
    Listing queues ...
    q1      0
    q2      3
    ...done.

As you can see recv.py receives messages from q1 only. To receive messages from q2, we need to run
recv.py once more.

    C:\pika\one-queue-issue>python recv.py 2> recvlog2.txt
    body: q2: message 0
    body: q2: message 1
    body: q2: message 2

    C:\pika\one-queue-issue>rabbitmqctl list_queues
    Listing queues ...
    q1      0
    q2      0
    ...done.

See <https://gist.github.com/susam/7173375> for the logs generated by the above scripts.

This issue does not occur if we first run the consumer program (recv.py), and then start the producer program (send.py). In this case, the consumer receives messages from both queues.

**Comment by maziyarpanahi** [comment]
I started to see if I can listen to multiple queues and I notice the exact same situation. If there are messages in those queues only one will be used and the other not at all. But if the publisher starts after the receiver has been running it's all good.

Is there any new information about this?

Thanks

**Comment by vitaly-krugl** [comment]
Thank you. This is fixed in pika 0.10.0:

```
In [5]: %paste
import pika
import logging

logging.basicConfig(level=logging.DEBUG)

parameters = pika.ConnectionParameters('localhost')
connection = pika.BlockingConnection(parameters)
channel = connection.channel()

# Declare queues for the consumer
channel.queue_declare(queue='q1')
channel.queue_declare(queue='q2')

# Publish messages
for i in range(3):
    message = 'message ' + str(i)
    channel.basic_publish(exchange='',
                          routing_key='q1',
                          body='q1: ' + message)
    channel.basic_publish(exchange='',
                          routing_key='q2',
                          body='q2: ' + message)
    print 'Published', message, 'with routing keys q1 and q2'

connection.close()

## -- End pasted text --
Published message 0 with routing keys q1 and q2
Published message 1 with routing keys q1 and q2
Published message 2 with routing keys q1 and q2

In [6]: %paste
import pika
import logging

logging.basicConfig(level=logging.DEBUG)

def print_data(channel, method, properties, body):
    print 'body:', body
    channel.basic_ack(delivery_tag = method.delivery_tag)

parameters = pika.ConnectionParameters('localhost')
connection = pika.BlockingConnection(parameters)
channel = connection.channel()

channel.queue_declare(queue='q1')
channel.queue_declare(queue='q2')

channel.basic_consume(print_data, queue='q1')
channel.basic_consume(print_data, queue='q2')

channel.start_consuming()
connection.close()

## -- End pasted text --
body: q1: message 0
body: q1: message 1
body: q1: message 2
body: q2: message 0
body: q2: message 1
body: q2: message 2
```
